RLBotFramework/utils/class_importer.py

```diff
@@ -1,75 +1,73 @@
-from RLBotFramework.utils.logging_utils import log_warn
 
 
-def is_extends_base_class(module, base_class):
-    super_classes = module.__bases__
-    for super_class in super_classes:
-        if super_class == base_class:
-            return True
-    return False
+class AgentLoadData:
+    def __init__(self, module_spec, base_class):
+        self.module_spec = module_spec
+        self.base_class = base_class
+        self.agent_class = None
+        self.agent_module = None
+        self.reload()
+
+    def reload(self):
+        self.agent_module = importlib.util.module_from_spec(self.module_spec)
+        sys.modules[self.module_spec.name] = self.agent_module
+        self.module_spec.loader.exec_module(self.agent_module)
+        self.agent_class = extract_agent_class(self.agent_module, self.base_class)
 
 
-def import_agent(module_name):
+def import_agent(python_file):
     """
     Imports the first class that extends BaseAgent.
 
-    :param module_name: A string formatted like a normal python import
+    :param python_file: The absolute path of the bot's main python file
     :return: The agent requested or BaseAgent if there are any problems.
     """
-    return import_class_with_base(module_name, BaseAgent)
+    return import_class_with_base(python_file, BaseAgent)
 
 
-def import_class_with_base(module_name, base_class):
+def import_class_with_base(python_file, base_class):
     """
     Imports the first class that extends base_class.
 
-    :param module_name: A string formatted like a normal python import
-    :param base_class: The class that we look for the extension for also is returned if no matching module is found
+    :param python_file: The absolute path of the bot's main python file
+    :param base_class: The class that we look for the extension for
     :return: The agent requested or BaseAgent if there are any problems.
     """
-    try:
-        module = importlib.import_module(module_name)
-        agent_class = [agent[1] for agent in inspect.getmembers(module, inspect.isclass)
-                       if issubclass(agent[1], base_class) and agent[1].__module__ == module_name]
+    dir_name = os.path.dirname(python_file)
+    package_name = str(uuid.uuid1())  # Pick a random package name so that we don't have to worry about collisions.
 
-        agent = agent_class[0]
-        # grabs only the first one
-        return agent
-    except ModuleNotFoundError as e:
-        log_warn('ModuleNotFoundError: %s\nsub module %s not found using %s instead', [str(e), str(module_name),
-                                                                                       str(base_class)])
-    except Exception as e:
-        log_warn('Error: %s\n%s not found using %s instead', [str(e), str(module_name), str(base_class)])
+    init_name = os.path.join(dir_name, "__init__.py")
+    if not os.path.exists(init_name):
+        with open(init_name, 'w'):  # Create an empty __init__.py file if it doesn't exist

```

Comment:
you do not need the with
you should be able to just do
`open(init_name, 'w')`

Reply:
I believe `with` has the advantage of closing the resource when the block exits.